File: python/tests/test_rect_window_and_renyi2_degenerate.py
"""Regression tests for two v3 out-of-band fixes.

#20  weight_events' rectangular window uses a half-open support
     [c - W/2, c + W/2): a regular pulse grid yields exactly N pulses for
     full support N*IOI at every N (a closed interval over-counts even
     widths and can leave a between-pulse centre empty).

#21  renyi2 returns NaN for a zero-mass density (e.g. an out-of-support
     windowed sweep centre) rather than raising.
"""
import numpy as np
import pytest

import mpt
from mpt._tensor.preprocessing import weight_events


def _pulse_count(centre, width, n=9):
    """Number of pulses retained by a rectangular window on a unit grid."""
    pitch = np.arange(n, dtype=float).reshape(1, n)
    time = np.arange(n, dtype=float).reshape(1, n)
    _, w_out, _ = weight_events(
        [pitch, time], None, input_attr=1, target_attr=0,
        centre=centre, shape=1.0, width=width, drop_input_attr=False,
    )
    return int(np.count_nonzero(w_out[0][0]))


class TestRectWindowHalfOpen:
    """#20: half-open [c - W/2, c + W/2) rectangular support."""

    @pytest.mark.parametrize("width,expected",
                             [(1, 1), (2, 2), (3, 3), (4, 4), (5, 5)])
    def test_on_pulse_width_to_count(self, width, expected):
        # Centre on a pulse (IOI = 1). A closed interval would collapse
        # widths 1..5 to pulse counts 1, 3, 3, 5, 5.
        assert _pulse_count(4.0, float(width)) == expected

    def test_between_pulse_centre_not_empty(self):
        # Midway between pulses 3 and 4; width 1 keeps the lower-edge
        # pulse (count 1), not zero (the closed/FP degeneracy) or two.
        assert _pulse_count(3.5, 1.0) == 1

    def test_lower_edge_in_upper_edge_out(self):
        # width 2 at centre 4 -> [3, 5): pulses 3 and 4; pulse 5 excluded.
        pitch = np.arange(9, dtype=float).reshape(1, 9)
        time = np.arange(9, dtype=float).reshape(1, 9)
        _, w_out, _ = weight_events(
            [pitch, time], None, input_attr=1, target_attr=0,
            centre=4.0, shape=1.0, width=2.0, drop_input_attr=False,
        )
        factor = w_out[0][0]
        assert factor[3] > 0 and factor[4] > 0   # lower edge + interior
        assert factor[5] == 0.0                  # upper edge excluded

    def test_fractional_grid_mapping(self):
        # Non-integer IOI: pulses spaced 0.25, width 4*IOI = 1.0 -> 4 pulses.
        n = 9
        pitch = np.arange(n, dtype=float).reshape(1, n)
        time = (0.25 * np.arange(n, dtype=float)).reshape(1, n)
        _, w_out, _ = weight_events(
            [pitch, time], None, input_attr=1, target_attr=0,
            centre=time[0, 4], shape=1.0, width=1.0, drop_input_attr=False,
        )
        assert int(np.count_nonzero(w_out[0][0])) == 4

    def test_gaussian_shape_unaffected(self):
        pitch = np.arange(5, dtype=float).reshape(1, 5)
        time = np.arange(5, dtype=float).reshape(1, 5)
        _, w_out, _ = weight_events(
            [pitch, time], None, input_attr=1, target_attr=0,
            centre=2.0, shape=0.0, sd=1.0, drop_input_attr=False,
        )
        factor = w_out[0][0]
        assert factor[2] == pytest.approx(1.0)   # peak at centre
        assert np.all(factor > 0)                # no hard edge


class TestRenyi2ZeroMassNaN:
    """#21: zero-mass density -> NaN from renyi2, not a raise."""

    def test_single_multiset_zero_weight_returns_nan(self):
        v = mpt.entropy_exp_tens([0., 4., 7.], [0., 0., 0.], 1.0, 1,
                                 False, False, 0.0, method="renyi2",
                                 verbose=False)
        assert np.isnan(v)

    def test_ma_out_of_support_window_returns_nan(self):
        pitch = np.array([[60., 62., 64.]])
        time = np.array([[0., 1., 2.]])
        # Rectangular window centred far from every event -> zero mass.
        pa, wa, sp = weight_events(
            [pitch, time], None, input_attr=1, target_attr=0,
            centre=100.0, shape=1.0, width=1.0, drop_input_attr=False,
        )
        dens = mpt.build_exp_tens(
            pa, wa, specs=sp, sigma=[1.0, 1.0],
            is_per=[False, False], period=[0., 0.], verbose=False,
        )
        v = mpt.entropy_exp_tens(dens, method="renyi2", verbose=False)
        assert np.isnan(v)

File: matlab/tests/test_rect_window_and_renyi2_degenerate.m
%% test_rect_window_and_renyi2_degenerate.m — two v3 out-of-band fixes
%
%  #20  weightEvents' rectangular window uses a half-open support
%       [c - W/2, c + W/2): a regular pulse grid yields exactly N pulses
%       for full support N*IOI at every N. A closed interval over-counts
%       even widths (widths 1..5 collapse to pulse counts 1, 3, 3, 5, 5)
%       and, at a between-pulse centre, can drop both flanking edge
%       pulses and leave a zero-mass density.
%
%  #21  entropyExpTens with method='renyi2' returns NaN for a zero-mass
%       density (for instance a windowed sweep centre with no event in
%       support) rather than raising. Sweep callers already want NaN
%       outside the data.
%
%  Python twin: tests/test_rect_window_and_renyi2_degenerate.py, whose
%  eight checks this file mirrors one for one.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

% Unit pulse grid: pitch and time both run 0, 1, ..., 8. The window is
% driven by time (attribute 2) and written to pitch (attribute 1).
rwPitch9 = 0:8;
rwTime9  = 0:8;

% --- 1. On-pulse centre: width N keeps exactly N pulses ---------------
rwOk = true;
for rwW = 1:5
    [~, rwWc, ~] = unpackPreMaet(weightEvents({rwPitch9, rwTime9}, [], 2, 1, 4, 1, ...
                                'width', rwW, 'dropInputAttr', false));
    rwOk = rwOk && (nnz(rwWc{1}) == rwW);
end
results{end+1, 1} = 'rectWindow: width N keeps exactly N pulses (on-pulse centre)'; %#ok<*AGROW>
results{end, 2}   = rwOk;

% --- 2. Between-pulse centre is not empty -----------------------------
%  Midway between pulses 3 and 4; width 1 keeps the lower-edge pulse,
%  not zero (the closed-interval degeneracy) and not two.
[~, rwWb, ~] = unpackPreMaet(weightEvents({rwPitch9, rwTime9}, [], 2, 1, 3.5, 1, ...
                            'width', 1, 'dropInputAttr', false));
results{end+1, 1} = 'rectWindow: between-pulse centre keeps one pulse';
results{end, 2}   = nnz(rwWb{1}) == 1;

% --- 3. Lower edge included, upper edge excluded ----------------------
%  width 2 at centre 4 spans [3, 5): pulses 3 and 4, not pulse 5.
[~, rwWe, ~] = unpackPreMaet(weightEvents({rwPitch9, rwTime9}, [], 2, 1, 4, 1, ...
                            'width', 2, 'dropInputAttr', false));
rwFactor = rwWe{1};
results{end+1, 1} = 'rectWindow: lower edge included, upper edge excluded';
results{end, 2}   = rwFactor(4) > 0 && rwFactor(5) > 0 && rwFactor(6) == 0;

% --- 4. Non-integer inter-onset interval ------------------------------
%  Pulses spaced 0.25; full support 1.0 = 4 * IOI keeps 4 pulses.
rwPitch = 0:8;
rwTime  = 0.25 * (0:8);
[~, rwW4, ~] = unpackPreMaet(weightEvents({rwPitch, rwTime}, [], 2, 1, rwTime(5), 1, ...
                            'width', 1, 'dropInputAttr', false));
results{end+1, 1} = 'rectWindow: fractional IOI grid keeps width/IOI pulses';
results{end, 2}   = nnz(rwW4{1}) == 4;

% --- 5. The Gaussian shape is untouched by the half-open rule ---------
rwPitch5 = 0:4;
rwTime5  = 0:4;
[~, rwW5, ~] = unpackPreMaet(weightEvents({rwPitch5, rwTime5}, [], 2, 1, 2, 0, ...
                            'sd', 1, 'dropInputAttr', false));
rwF5 = rwW5{1};
results{end+1, 1} = 'rectWindow: Gaussian shape peak-normalised with no hard edge';
results{end, 2}   = abs(rwF5(3) - 1) < 1e-12 && all(rwF5 > 0);

% --- 6. renyi2 on an all-zero-weight single multiset returns NaN ------
rwH1 = entropyExpTens([0 4 7], [0 0 0], 1, 1, false, false, 0, ...
                      'method', 'renyi2', 'verbose', false);
results{end+1, 1} = 'renyi2: all-zero weights return NaN, not an error';
results{end, 2}   = isnan(rwH1);

% --- 7. renyi2 on an out-of-support windowed density returns NaN ------
rwPitch7 = [60 62 64];
rwTime7  = [0 1 2];
[rwPa, rwWa, rwSp] = unpackPreMaet(weightEvents({rwPitch7, rwTime7}, [], 2, 1, 100, 1, ...
                                  'width', 1, 'dropInputAttr', false));
rwDens = buildExpTens(rwPa, rwWa, 'specs', rwSp, 'sigma', [1 1], ...
                      'isPer', [false false], 'period', [0 0], ...
                      'verbose', false);
rwH2 = entropyExpTens(rwDens, 'method', 'renyi2', 'verbose', false);
results{end+1, 1} = 'renyi2: out-of-support window returns NaN, not an error';
results{end, 2}   = isnan(rwH2);

% --- 8. A healthy density still returns a finite value ----------------
%  Guards against the NaN paths above swallowing every case.
rwH3 = entropyExpTens([0 4 7], [], 30, 2, false, false, 0, ...
                      'method', 'renyi2', 'verbose', false);
results{end+1, 1} = 'renyi2: a healthy density still returns a finite value';
results{end, 2}   = isfinite(rwH3);

if standalone
    nPass = sum(cellfun(@(x) isequal(x, true), results(:,2)));
    nFail = numel(results(:,1)) - nPass;
    fprintf('\n=== test_rect_window_and_renyi2_degenerate: %d passed, %d failed (of %d) ===\n', ...
        nPass, nFail, numel(results(:,1)));
    clear cleanupDefaults
    if nFail > 0
        for ii = 1:size(results, 1)
            if ~isequal(results{ii, 2}, true)
                fprintf('  FAIL  %s\n', results{ii, 1});
            end
        end
    end
end
